### PR TITLE
adding support for qualifier on bootstrap roles

### PR DESCRIPTION
**Pull request by dgraeber**
*Issue #, if available:*
#354 

*Description of changes:*
Adding support for qualifiers on bootstrap roles

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: docs/source/bootstrapping.md [hunk 1]
```diff
@@ -60,5 +63,12 @@ You must pass in the `toolchain-account` field so a trust-relationship can be se
 
 The `permission-boundary` filed allows you to attach a policy to the role to act as a [permissions boundary](https://docs.aws.amazon.com/IAM/latest/UserGuide/access_policies_boundaries.html)
 
+
+## Qualifiers for Toolchain Role and Target Roles
+We have added suppprt for the use of a qualifier for the toolchain role and the target role(s).  This is to help segregate target deployments when using a single CICD account as the toolchain account across multiple deployment schemes (ex. DEV, INT, PROD).  Most cases do not need this support, but unique deployment scenarios may require it.
```
Comment (srinivasreddych) [inline docs/source/bootstrapping.md:68]: Need to re-write it. Any enterprise which follows multi account deployments might need to adopt similar deployment plan(s)

File: seedfarmer/utils.py [hunk 2]
```diff
@@ -126,3 +126,25 @@ def generate_codebuild_url(account_id: str, region: str, codebuild_id: str) -> s
...
+    name = f"seedfarmer-{project_name}-deployment-role"
+    return f"{name}-{qualifier}" if qualifier else name
+
+
+def get_deployment_role_arn(deployment_account_id: str, project_name: str, qualifier: Optional[str] = None) -> str:
+    return f"arn:aws:iam::{deployment_account_id}:role/{get_deployment_role_name(project_name,qualifier)}"
+
+
+def valid_qualifier(qualifer: str) -> bool:
+    return True if ((len(qualifer) <= 6) and qualifer.isalnum()) else False
```
Comment (srinivasreddych) [inline seedfarmer/utils.py:150]: Are you being flexible for len<6?

File: docs/source/bootstrapping.md [hunk 3]
```diff
@@ -46,12 +47,14 @@ Options:
...
                                   Role  [required]
-  -b, --permissions-boundary TEXT  ARN of a Managed Policy to set as the
+  -b, --permissions-boundary TEXT
+                                  ARN of a Managed Policy to set as the
                                   Permission Boundary on the Toolchain Role
   --synth / --no-synth            Synthesize a CFN template only...do not
                                   deploy  [default: no-synth]
   --profile TEXT                  The AWS profile to initiate a session
   --region TEXT                   AWS region to use
+  --qualifier TEXT                A qualifier to append to toolchain role
```
Comment (srinivasreddych) [inline docs/source/bootstrapping.md:57]: Are you appending qualifier only to toolchain? Not to the relevant deployment roles?

File: docs/source/bootstrapping.md [hunk 4]
```diff
@@ -14,14 +14,16 @@ Options:
...
+  -b, --permissions-boundary TEXT
+                                  ARN of a Managed Policy to set as the
                                   Permission Boundary on the Toolchain Role
   --as-target / --not-as-target   Optionally also bootstrap the account as a
                                   Target account  [default: not-as-target]
   --synth / --no-synth            Synthesize a CFN template only...do not
                                   deploy  [default: no-synth]
   --profile TEXT                  The AWS profile to initiate a session
   --region TEXT                   AWS region to use
+  --qualifier TEXT                A qualifier to append to toolchain role
```
[the remark this answers is not in the document]
Comment (srinivasreddych) [reply docs/source/bootstrapping.md:26]: Also its good to mention the length expectations here IMO

File: seedfarmer/__main__.py [hunk 5]
```diff
@@ -57,6 +57,12 @@ def version() -> None:
     help="The AWS region to use for boto3.Sessions",
     required=False,
 )
+@click.option(
+    "--qualifier",
+    default=None,
+    help="A qualifier to append to toolchain role",
```
Comment (srinivasreddych) [inline seedfarmer/__main__.py:63]: Are you appending qualifier only to toolchain? Not to the relevant deployment roles? If to both, the description should be generic

File: test/unit-test/test_commands_bootstrap.py [hunk 6]
```diff
@@ -64,7 +64,9 @@ def session_manager(sts_client):
 def test_deploy_template(session_manager, mocker):
     mocker.patch("seedfarmer.commands._bootstrap_commands.role_deploy_status",return_value={"RoleName":"BLAHBLAH"})
     mocker.patch("seedfarmer.commands._bootstrap_commands.cs_services.cfn.deploy_template",return_value=None)
-    template = bc.get_toolchain_template(project_name="myapp",principal_arn=['arn:aws:iam::123456789012:role/AdminRole']
+    template = bc.get_toolchain_template(project_name="myapp",
+                                         role_name="seedfarmer-test-toolchain-role",
+                                         principal_arn=['arn:aws:iam::123456789012:role/AdminRole']
```
Comment (chamcca) [inline test/unit-test/test_commands_bootstrap.py:69]: this formatting looks off.

**Review by chamcca — COMMENTED** [review]
i don't see tests using the new qualifier functionality

File: docs/source/bootstrapping.md (hunk 1 above)
Comment (dgraeber) [reply docs/source/bootstrapping.md:68]: what would you suggest?  I can remove the line 'Most cases do not need this support'

File: seedfarmer/utils.py (hunk 2 above)
Comment (dgraeber) [reply seedfarmer/utils.py:150]: ?  6 is the max the length the qualifier can be.

File: docs/source/bootstrapping.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (dgraeber) [reply docs/source/bootstrapping.md:26]: bah..that comment is off...

File: docs/source/bootstrapping.md (hunk 3 above)
Comment (dgraeber) [reply docs/source/bootstrapping.md:57]: it is being appended to toolchain role...that text is off

File: seedfarmer/__main__.py (hunk 5 above)
Comment (dgraeber) [reply seedfarmer/__main__.py:63]: it is being appended to toolchain role...that text is off

File: seedfarmer/utils.py (hunk 2 above)
Comment (srinivasreddych) [reply seedfarmer/utils.py:150]: To clarify more, are we being flexible for variable length between 0-6? Didnt we decide it would be 6 only?

**Comment by srinivasreddych** [comment]
Overall, there are lot of places, where the description for qualifier needs to be slightly adjusted, else a reader might think the qual is only being added to `tc` role but not to the `dep role`

File: docs/source/bootstrapping.md (hunk 1 above)
Comment (srinivasreddych) [reply docs/source/bootstrapping.md:68]: We have added support for the use of a qualifier for naming the Toolchain role and Target account's deployment role(s). This is to help segregate and standardize target deployments when using a multi account structure which has a central shared services (CICD account) as the toolchain account performing deployments across relevant environments (ex. DEV, INT, PROD). `qualifier` can be mainly used if you want to restrict the level of access/action a dev/tester/support team can perform on any target given environment

File: test/unit-test/test_commands_bootstrap.py (hunk 6 above)
Comment (chamcca) [reply test/unit-test/test_commands_bootstrap.py:69]: black wouldn't format it aligned to the `(`.

Comment (dgraeber) [reply test/unit-test/test_commands_bootstrap.py]: hmmm....looking into it

File: seedfarmer/utils.py (hunk 2 above)
Comment (dgraeber) [reply seedfarmer/utils.py:150]: I think we are being more flexible.... ;)

File: test/unit-test/test_commands_bootstrap.py (hunk 6 above)
Comment (dgraeber) [reply test/unit-test/test_commands_bootstrap.py]: Formatted all tests ;).

**Review by srinivasreddych — APPROVED** [review]

File: docs/source/bootstrapping.md [hunk 7]
```diff
@@ -60,5 +63,12 @@ You must pass in the `toolchain-account` field so a trust-relationship can be se
 
 The `permission-boundary` filed allows you to attach a policy to the role to act as a [permissions boundary](https://docs.aws.amazon.com/IAM/latest/UserGuide/access_policies_boundaries.html)
 
+
+## Qualifiers for Toolchain Role and Target Roles
+We have added suppprt for the use of a qualifier for the toolchain role and the target account deployment role(s).  This is to help segregate target deployment when using a multi-account structure which has a central shared services (CICD account) as the toolchain account performing deployments across relevant environments (ex. DEV, INT, PROD). A `qualifier` can be used if you want to restrict the level of access/action a dev/tester/support team can perform on any target given environment.
```
[the remark this answers is not in the document]
Comment (dgraeber) [reply docs/source/bootstrapping.md:68]: very minor ;)

**Pull request merged**